File: Javascript/backend_nest/src/domain/entities/Author.ts
interface AuthorInput {
  id: number;
  firstName: string;
  lastName: string;
  birthDate: Date;
  nationality: string;
  deathDate: Date | null;
  biography?: string;
  books?: number[];
}

export class Author {
  private static readonly ALLOWED_NATIONALITIES = new Set([
    'Français',
    'Anglais',
    'Américain',
    'Espagnol',
    'Allemand',
    'Colombien',
  ]);

  private _id: number;
  private _firstName: string;
  private _lastName: string;
  private _birthDate: Date;
  private _nationality: string;
  private _deathDate: Date | null;
  private _biography?: string;
  private _books: number[];

  constructor(input: AuthorInput) {
    const { id, firstName, lastName, birthDate, deathDate, nationality, biography, books } = input;

    this._id = id;
    this._firstName = firstName;
    this._lastName = lastName;
    this._birthDate = birthDate;
    this._nationality = nationality;
    this._deathDate = deathDate;
    this._biography = biography;
    this._books = books;
  }

  static create(input: AuthorInput): Author {
    const { nationality, deathDate, birthDate } = input;
    if (!Author.ALLOWED_NATIONALITIES.has(nationality)) {
      throw new Error(
        `Nationalité '${nationality}' non reconnue. Les nationalités acceptées sont : ${Array.from(
          Author.ALLOWED_NATIONALITIES
        ).join(', ')}`
      );
    }

    if (deathDate && birthDate >= deathDate) {
      throw new Error('La date de naissance doit être antérieure à la date de décès.');
    }

    return new Author(input);
  }

  setId(id: number): void {
    this._id = id;
  }

  get id(): number {
    return this._id;
  }

  get firstName(): string {
    return this._firstName;
  }

  get lastName(): string {
    return this._lastName;
  }

  get birthDate(): Date {
    return this._birthDate;
  }

  get nationality(): string {
    return this._nationality;
  }

  get deathDate(): Date | null {
    return this._deathDate;
  }

  get biography(): string | undefined {
    return this._biography;
  }

  get books(): number[] {
    return this._books;
  }
}

File: Python/backend_fastapi/domain/entities/Author.py
from datetime import date
from typing import List, Optional

class Author:
    ALLOWED_NATIONALITIES = {"Français", "Anglais", "Américain", "Espagnol", "Allemand"}

    def __init__(self, id: int, first_name: str, last_name: str, birth_date: date, nationality: Optional[str] = None,
                 death_date: Optional[date] = None, biography: Optional[str] = None, books: Optional[List[int]] = None):
        self._id = id
        self._first_name = first_name
        self._last_name = last_name
        self._birth_date = birth_date
        self._nationality = nationality
        self._death_date = death_date
        self._biography = biography
        self._books = books or []

    @staticmethod
    def create(id: int, first_name: str, last_name: str, birth_date: date, nationality: Optional[str] = None,
               death_date: Optional[date] = None, biography: Optional[str] = None, books: Optional[List[int]] = None):

        if nationality and nationality not in Author.ALLOWED_NATIONALITIES:
            raise ValueError(f"Nationalité '{nationality}' non reconnue. Les nationalités acceptées sont : {Author.ALLOWED_NATIONALITIES}")
        
        if death_date and birth_date >= death_date:
            raise ValueError("La date de naissance doit être antérieure à la date de décès.")
        
        return Author(id, first_name, last_name, birth_date, nationality, death_date, biography, books)

    @property
    def id(self) -> int:
        return self._id

    @property
    def first_name(self) -> str:
        return self._first_name
    
    @property
    def last_name(self) -> str:
        return self._last_name

    @property
    def birth_date(self) -> date:
        return self._birth_date

    @property
    def nationality(self) -> Optional[str]:
        return self._nationality

    @property
    def death_date(self) -> Optional[date]:
        return self._death_date

    @property
    def biography(self) -> Optional[str]:
        return self._biography
    
    @property
    def books(self) -> List[int]:
        return self._books
    
